traffic_ops/traffic_ops_golang/server/servers_server_capability.go

```diff
@@ -453,77 +443,185 @@ func AssignMultipleServerCapabilities(w http.ResponseWriter, r *http.Request) {
+	if len(mssc.ServerIDs) >= 1 {
+		errCode, userErr, sysErr = checkExistingServer(tx, mssc.ServerIDs, inf.User.UserName)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+			return
+		}
+	}
 
-	mscQuery := `WITH inserted AS (
-		INSERT INTO server_server_capability
-		SELECT "server_capability", $2
-		FROM UNNEST($1::text[]) AS tmp("server_capability")
-		RETURNING server_capability
-		)
-		SELECT ARRAY_AGG(server_capability)
-		FROM (
-			SELECT server_capability
-			FROM inserted
-		) AS returned(server_capability)`
+	//Delete existing rows from server_server_capability for a given server or for a given capability
+	const delQuery = `DELETE FROM server_server_capability ssc WHERE `
+	var dq string
+	var alerts tc.Alerts
+	var result sql.Result
+	var err error
+	if mssc.PageType == "sc" {
+		dq = delQuery + `ssc.server_capability=$1`
+		if len(mssc.ServerIDs) == 1 {
+			dq = dq + ` AND ssc.server=$2`
+			result, err = tx.Exec(dq, mssc.ServerCapabilities[0], mssc.ServerIDs[0])
+		} else {
+			result, err = tx.Exec(dq, mssc.ServerCapabilities[0])
+		}
+	} else {
+		dq = delQuery + `ssc.server=$1`
+		if len(mssc.ServerCapabilities) == 1 {
+			dq = dq + ` AND ssc.server_capability=$2`
+			result, err = tx.Exec(dq, mssc.ServerIDs[0], mssc.ServerCapabilities[0])
+		} else {
+			result, err = tx.Exec(dq, mssc.ServerIDs[0])
+		}
+	}
 
-	err = tx.QueryRow(mscQuery, pq.Array(msc.ServerCapabilities), msc.ServerID).Scan(pq.Array(&multipleServerCapabilities))
 	if err != nil {
 		useErr, sysErr, statusCode := api.ParseDBError(err)
 		api.HandleErr(w, r, tx, statusCode, useErr, sysErr)
 		return
 	}
-	msc.ServerCapabilities = multipleServerCapabilities
-	alerts := tc.CreateAlerts(tc.SuccessLevel, "Multiple Server Capabilities assigned to a server")
-	api.WriteAlertsObj(w, r, http.StatusOK, alerts, msc)
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, fmt.Errorf("no rows were deleted from server_server_capability table: %w", err), sysErr)
+		return
+	}
+	if rowsAffected >= 1 {
+		if mssc.PageType == "sc" {
+			alerts = tc.CreateAlerts(tc.SuccessLevel, "Removed Server(s) associated with a capability")
+		} else {
+			alerts = tc.CreateAlerts(tc.SuccessLevel, "Removed Server Capability(ies) associated with a server")
+		}
+	}
+
+	api.WriteAlertsObj(w, r, http.StatusOK, alerts, mssc)
 	return
 }
+
+// checkExistingServer checks server existence
+func checkExistingServer(tx *sql.Tx, sidList []int64, uName string) (int, error, error) {
+	for _, sid := range sidList {
+		_, exists, err := dbhelpers.GetServerNameFromID(tx, sid)
+		if err != nil {
+			return http.StatusInternalServerError, nil, err
+		}
+		if !exists {
+			userErr := fmt.Errorf("server %d does not exist", sid)
+			return http.StatusNotFound, userErr, nil
+		}
+
+		cdnName, err := dbhelpers.GetCDNNameFromServerID(tx, sid)
```

Comment:
Can we get the CDN name from the server ID directly? That saves us a DB query.

Reply:
Unfortunately we cannot.

Reply:
How about something like this?
`select c.name from cdn c join server s on s.cdn_id = c.id where s.id=<server ID>`

Reply:
Nvm, this might be scope creep, so I'll create a separate issue for it, wherein, we just perform one query to check if the server exists or not, and also get the CDN name from the server ID.